File: build_user/app/tmp/kapt3/stubs/debug/com/assistx/monitor/service/MonitoringService.java
package com.assistx.monitor.service;

@kotlin.Metadata(mv = {1, 9, 0}, k = 1, xi = 48, d1 = {"\u0000R\n\u0002\u0018\u0002\n\u0002\u0018\u0002\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0002\u0010\u000e\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\t\n\u0002\b\u0002\n\u0002\u0018\u0002\n\u0000\n\u0002\u0018\u0002\n\u0000\n\u0002\u0010\u0002\n\u0002\b\u0002\n\u0002\u0010\b\n\u0002\b\u0004\n\u0002\u0010 \n\u0002\b\u0006\u0018\u0000  2\u00020\u0001:\u0001 B\u0005\u00a2\u0006\u0002\u0010\u0002J\u000e\u0010\u000b\u001a\u00020\fH\u0082@\u00a2\u0006\u0002\u0010\rJ\u0012\u0010\u000e\u001a\u0004\u0018\u00010\u000f2\u0006\u0010\u0010\u001a\u00020\u0011H\u0016J\b\u0010\u0012\u001a\u00020\u0013H\u0016J\b\u0010\u0014\u001a\u00020\u0013H\u0016J\"\u0010\u0015\u001a\u00020\u00162\b\u0010\u0010\u001a\u0004\u0018\u00010\u00112\u0006\u0010\u0017\u001a\u00020\u00162\u0006\u0010\u0018\u001a\u00020\u0016H\u0016J\u0018\u0010\u0019\u001a\u00020\u00132\u000e\u0010\u001a\u001a\n\u0012\u0004\u0012\u00020\b\u0018\u00010\u001bH\u0002J\b\u0010\u001c\u001a\u00020\u0013H\u0002J\b\u0010\u001d\u001a\u00020\u0013H\u0002J\b\u0010\u001e\u001a\u00020\u0013H\u0002J\b\u0010\u001f\u001a\u00020\u0013H\u0002R\u0010\u0010\u0003\u001a\u0004\u0018\u00010\u0004X\u0082\u000e\u00a2\u0006\u0002\n\u0000R\u001a\u0010\u0005\u001a\u000e\u0012\u0004\u0012\u00020\u0007\u0012\u0004\u0012\u00020\b0\u0006X\u0082\u0004\u00a2\u0006\u0002\n\u0000R\u000e\u0010\t\u001a\u00020\nX\u0082.\u00a2\u0006\u0002\n\u0000\u00a8\u0006!"}, d2 = {"Lcom/assistx/monitor/service/MonitoringService;", "Landroidx/lifecycle/LifecycleService;", "()V", "pollingJob", "Lkotlinx/coroutines/Job;", "previousStates", "Ljava/util/concurrent/ConcurrentHashMap;", "", "Lcom/assistx/monitor/data/model/PcDevice;", "sseManager", "Lcom/assistx/monitor/sync/SSEManager;", "getPollIntervalMs", "", "(Lkotlin/coroutines/Continuation;)Ljava/lang/Object;", "onBind", "Landroid/os/IBinder;", "intent", "Landroid/content/Intent;", "onCreate", "", "onDestroy", "onStartCommand", "", "flags", "startId", "processDeviceUpdate", "devices", "", "startForeground", "startMonitoring", "startPollingFallback", "stopPollingFallback", "Companion", "app_debug"})
public final class MonitoringService extends androidx.lifecycle.LifecycleService {
    public static final int NOTIFICATION_ID = 1001;
    @org.jetbrains.annotations.NotNull
    public static final java.lang.String ACTION_STOP = "com.assistx.monitor.STOP_SERVICE";
    private static final long DEFAULT_POLL_INTERVAL_MS = 10000L;
    private com.assistx.monitor.sync.SSEManager sseManager;
    @org.jetbrains.annotations.Nullable
    private kotlinx.coroutines.Job pollingJob;
    @org.jetbrains.annotations.NotNull
    private final java.util.concurrent.ConcurrentHashMap<java.lang.String, com.assistx.monitor.data.model.PcDevice> previousStates = null;
    @org.jetbrains.annotations.NotNull
    public static final com.assistx.monitor.service.MonitoringService.Companion Companion = null;
    
    public MonitoringService() {
        super();
    }
    
    @java.lang.Override
    public void onCreate() {
    }
    
    @java.lang.Override
    public int onStartCommand(@org.jetbrains.annotations.Nullable
    android.content.Intent intent, int flags, int startId) {
        return 0;
    }
    
    @java.lang.Override
    @org.jetbrains.annotations.Nullable
    public android.os.IBinder onBind(@org.jetbrains.annotations.NotNull
    android.content.Intent intent) {
        return null;
    }
    
    private final void startForeground() {
    }
    
    private final void startMonitoring() {
    }
    
    private final void startPollingFallback() {
    }
    
    private final void stopPollingFallback() {
    }
    
    /**
     * Baca poll interval dari preferences secara aman.
     * Fallback ke default jika gagal.
     */
    private final java.lang.Object getPollIntervalMs(kotlin.coroutines.Continuation<? super java.lang.Long> $completion) {
        return null;
    }
    
    private final void processDeviceUpdate(java.util.List<com.assistx.monitor.data.model.PcDevice> devices) {
    }
    
    @java.lang.Override
    public void onDestroy() {
    }
    
    @kotlin.Metadata(mv = {1, 9, 0}, k = 1, xi = 48, d1 = {"\u0000\u001e\n\u0002\u0018\u0002\n\u0002\u0010\u0000\n\u0002\b\u0002\n\u0002\u0010\u000e\n\u0000\n\u0002\u0010\t\n\u0000\n\u0002\u0010\b\n\u0000\b\u0086\u0003\u0018\u00002\u00020\u0001B\u0007\b\u0002\u00a2\u0006\u0002\u0010\u0002R\u000e\u0010\u0003\u001a\u00020\u0004X\u0086T\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0005\u001a\u00020\u0006X\u0082T\u00a2\u0006\u0002\n\u0000R\u000e\u0010\u0007\u001a\u00020\bX\u0086T\u00a2\u0006\u0002\n\u0000\u00a8\u0006\t"}, d2 = {"Lcom/assistx/monitor/service/MonitoringService$Companion;", "", "()V", "ACTION_STOP", "", "DEFAULT_POLL_INTERVAL_MS", "", "NOTIFICATION_ID", "", "app_debug"})
    public static final class Companion {
        
        private Companion() {
            super();
        }
    }
}

File: app/src/main/java/com/assistx/monitor/service/MonitoringService.kt
package com.assistx.monitor.service

import android.app.PendingIntent
import android.content.Intent
import android.os.IBinder
import androidx.core.app.NotificationCompat
import androidx.lifecycle.LifecycleService
import androidx.lifecycle.lifecycleScope
import com.assistx.monitor.AssistXApplication
import com.assistx.monitor.R
import com.assistx.monitor.data.model.PcDevice
import com.assistx.monitor.data.repository.DeviceRepository
import com.assistx.monitor.network.ApiClient
import com.assistx.monitor.sync.SSEManager
import com.assistx.monitor.sync.SseState
import com.assistx.monitor.ui.MainActivity
import kotlinx.coroutines.Job
import kotlinx.coroutines.delay
import kotlinx.coroutines.flow.first
import kotlinx.coroutines.isActive
import kotlinx.coroutines.launch
import java.util.concurrent.ConcurrentHashMap

class MonitoringService : LifecycleService() {

    companion object {
        const val NOTIFICATION_ID = 1001
        const val ACTION_STOP = "com.assistx.monitor.STOP_SERVICE"
        private const val DEFAULT_POLL_INTERVAL_MS = 10_000L
    }

    private lateinit var sseManager: SSEManager
    private var pollingJob: Job? = null
    private val previousStates = ConcurrentHashMap<String, PcDevice>()

    override fun onCreate() {
        super.onCreate()
        sseManager = SSEManager()
        startForeground()
        startMonitoring()
    }

    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (intent?.action == ACTION_STOP) {
            stopSelf()
            return START_NOT_STICKY
        }
        return START_STICKY
    }

    override fun onBind(intent: Intent): IBinder? = null

    private fun startForeground() {
        val pendingIntent = PendingIntent.getActivity(
            this, 0, Intent(this, MainActivity::class.java),
            PendingIntent.FLAG_UPDATE_CURRENT or PendingIntent.FLAG_IMMUTABLE
        )
        val notification = NotificationCompat.Builder(this, AssistXApplication.CHANNEL_SERVICE)
            .setContentTitle("AssistX Monitoring")
            .setContentText("Memantau semua perangkat 24/7")
            .setSmallIcon(R.drawable.ic_monitor)
            .setContentIntent(pendingIntent)
            .setOngoing(true)
            .setSilent(true)
            .setPriority(NotificationCompat.PRIORITY_LOW)
            .build()
        startForeground(NOTIFICATION_ID, notification)
    }

    private fun startMonitoring() {
        // Initial fetch saat service start
        lifecycleScope.launch {
            try {
                val devices = ApiClient.getApiService().getClients()
                DeviceRepository.updateDevices(devices)
                processDeviceUpdate(devices)
            } catch (_: Exception) { }
        }

        // Collect SSE device updates
        lifecycleScope.launch {
            try {
                sseManager.deviceUpdates.collect { devices ->
                    try {
                        DeviceRepository.updateDevices(devices)
                        processDeviceUpdate(devices)
                    } catch (_: Exception) { }
                }
            } catch (_: Exception) { }
        }

        // Observe SSE connection state untuk fallback polling
        lifecycleScope.launch {
            try {
                sseManager.connectionState.collect { state ->
                    when (state) {
                        SseState.DISCONNECTED -> startPollingFallback()
                        SseState.CONNECTED -> stopPollingFallback()
                        SseState.CONNECTING -> { /* tunggu */ }
                    }
                }
            } catch (_: Exception) { }
        }

        // Mulai SSE
        try {
            sseManager.connect()
        } catch (_: Exception) {
            startPollingFallback()
        }
    }

    private fun startPollingFallback() {
        if (pollingJob?.isActive == true) return // sudah berjalan
        pollingJob = lifecycleScope.launch {
            val pollMs = getPollIntervalMs()
            while (isActive) {
                try {
                    val devices = ApiClient.getApiService().getClients()
                    DeviceRepository.updateDevices(devices)
                    processDeviceUpdate(devices)
                } catch (_: Exception) { }
                delay(pollMs)
            }
        }
    }

    private fun stopPollingFallback() {
        pollingJob?.cancel()
        pollingJob = null
    }

    /**
     * Baca poll interval dari preferences secara aman.
     * Fallback ke default jika gagal.
     */
    private suspend fun getPollIntervalMs(): Long {
        return try {
            val seconds = AssistXApplication.instance.preferencesManager.pollInterval.first()
            (seconds * 1000L).coerceAtLeast(3000L)
        } catch (_: Exception) {
            DEFAULT_POLL_INTERVAL_MS
        }
    }

    private fun processDeviceUpdate(devices: List<PcDevice>?) {
        if (devices.isNullOrEmpty()) return
        val alertManager = try {
            AlertManager.getInstance(this)
        } catch (_: Exception) {
            return
        }
        for (device in devices) {
            val id = device.id ?: continue
            val prev = previousStates[id]
            if (prev != null) {
                try {
                    alertManager.checkAndAlert(prev, device)
                } catch (_: Exception) { }
            }
            previousStates[id] = device
        }
    }

    override fun onDestroy() {
        sseManager.destroy()
        pollingJob?.cancel()
        super.onDestroy()
    }
}
